src/library/VaultLib.sol

```diff
@@ -0,0 +1,217 @@
+    }
+
+    function convertAssetToBase(address asset_, uint256 assets) public view returns (uint256) {
+        if (asset_ == address(0)) revert IVault.ZeroAddress();
+        uint256 rate = IProvider(getVaultStorage().provider).getRate(asset_);
+        return assets.mulDiv(rate, 10 ** (getAssetStorage().assets[asset_].decimals), Math.Rounding.Floor);
+    }
+
+    function convertBaseToAsset(address asset_, uint256 assets) public view returns (uint256) {
+        if (asset_ == address(0)) revert IVault.ZeroAddress();
+        uint256 rate = IProvider(getVaultStorage().provider).getRate(asset_);
+        return assets.mulDiv(10 ** (getAssetStorage().assets[asset_].decimals), rate, Math.Rounding.Floor);
+    }
+
+    function addTotalAssets(uint256 baseAssets) public {
+        IVault.VaultStorage storage vaultStorage = getVaultStorage();
+        if (!vaultStorage.alwaysComputeTotalAssets) {
+            vaultStorage.totalAssets += baseAssets;
+        }
+    }
+
+    function subTotalAssets(uint256 baseAssets) public {
+        IVault.VaultStorage storage vaultStorage = getVaultStorage();
+        if (!vaultStorage.alwaysComputeTotalAssets) {
+            vaultStorage.totalAssets -= baseAssets;
+        }
+    }
+
+    function convertToAssets(address asset_, uint256 shares, Math.Rounding rounding)
+        public
+        view
+        returns (uint256, uint256)
+    {
+        uint256 totalAssets = IVault(address(this)).totalAssets();
+        uint256 totalSupply = getERC20Storage().totalSupply;
+        uint256 baseAssets = shares.mulDiv(totalAssets + 1, totalSupply + 1, rounding);
+        uint256 assets = convertBaseToAsset(asset_, baseAssets);
+        return (assets, baseAssets);
+    }
+
+    function convertToShares(address asset_, uint256 assets, Math.Rounding rounding)
+        public
+        view
+        returns (uint256, uint256)
+    {
+        uint256 totalAssets = IVault(address(this)).totalAssets();
+        uint256 totalSupply = getERC20Storage().totalSupply;
+        uint256 baseAssets = convertAssetToBase(asset_, assets);
+        uint256 shares = baseAssets.mulDiv(totalSupply + 1, totalAssets + 1, rounding);
+        return (shares, baseAssets);
+    }
+
+    function setProcessorRule(address target, bytes4 functionSig, IVault.FunctionRule calldata rule) public {
+        getProcessorStorage().rules[target][functionSig] = rule;
+        emit IVault.SetProcessorRule(target, functionSig, rule);
+    }
+
+    function updateAsset(uint256 index, IVault.AssetUpdateFields calldata fields) public {
+        IVault.AssetStorage storage assetStorage = getAssetStorage();
+        if (index >= assetStorage.list.length) {
+            revert IVault.InvalidAsset(address(0));
+        }
+
+        address asset_ = assetStorage.list[index];
+        IVault.AssetParams storage assetParams = assetStorage.assets[asset_];
+        assetParams.active = fields.active;
+        emit IVault.UpdateAsset(index, asset_, fields);
+    }
+
```

Comment:
let's add functionality for `function deleteAsset`

make sensible choices like not allowing delete of asset at index 0 and if the balance of it is non-zero 

think of edge cases